misk/src/main/kotlin/misk/web/ServletHttpCall.kt

```diff
@@ -36,36 +36,55 @@ internal data class ServletHttpCall(
     upstreamResponse.addHeaders(headers)
   }
 
+  override fun requireTrailers() {
+    upstreamResponse.requireTrailers()
+  }
+
+  override fun setResponseTrailer(name: String, value: String) {
+    upstreamResponse.setTrailer(name, value)
+  }
+
   override fun takeRequestBody(): BufferedSource? {
     val result = requestBody
     requestBody = null
     return result
   }
 
+  override fun putRequestBody(requestBody: BufferedSource) {
+    check(this.requestBody == null) { "previous request body leaked; take it first" }
+    this.requestBody = requestBody
+  }
+
   override fun takeResponseBody(): BufferedSink? {
     val result = responseBody
     responseBody = null
     return result
   }
 
+  override fun putResponseBody(responseBody: BufferedSink) {
+    check(this.responseBody == null) { "previous response body leaked; take it first" }
+    this.responseBody = responseBody
+  }
+
   override fun takeWebSocket(): WebSocket? {
     val result = webSocket
     webSocket = null
     return result
   }
 
-  override fun withRequestBody(requestBody: BufferedSource) = copy(requestBody = requestBody)
-
-  override fun withResponseBody(responseBody: BufferedSink) = copy(responseBody = responseBody)
-
-  override fun withWebSocket(webSocket: WebSocket) = copy(webSocket = webSocket)
+  override fun putWebSocket(webSocket: WebSocket) {
+    check(this.webSocket == null) { "previous web socket leaked; take it first" }
+    this.webSocket = webSocket
+  }
 
   /** Adapts the underlying servlet call to send data to the client. */
   interface UpstreamResponse {
     var statusCode: Int
     val headers: Headers
     fun setHeader(name: String, value: String)
     fun addHeaders(headers: Headers)
+    fun requireTrailers()
+    fun setTrailer(name: String, value: String)
```

Comment:
again, `addTrailer`? The response can have many

Reply:
Yes!

Reply:
Actually, studying this, the implementation does a set and not an add. (You can have multiple values for a headers otherwise).

Reply:
hmmm, ok. Still confusing to me. `Map` uses `put` for that; it's more or less the same I guess